TestIPBlockWithExcept e2e test flaky on Kind

**Describe the bug**
I noticed a test failure earlier today after enabling the test on Kind. It went away when I ran the test a second time. We may want to keep an eye on this and see if it happens again. See the logs here: https://github.com/vmware-tanzu/antrea/runs/353194354.

**Additional context**
```
I'm copy-pasting the logs here in case they get deleted by Github:
2019/12/17 19:26:55 Test logs (if any) will be exported under the '/tmp/antrea-test-163908562' directory
2019/12/17 19:26:55 Collecting information about K8s cluster
2019/12/17 19:26:56 Pod network: '10.10.0.0/16'
2019/12/17 19:26:56 Num nodes: 2
--- FAIL: TestIPBlockWithExcept (45.60s)
##[error]    fixtures.go:27: Creating K8s clientset
##[error]    fixtures.go:33: Creating 'antrea-test' K8s Namespace
##[error]    fixtures.go:37: Applying Antrea YAML
##[error]    fixtures.go:41: Waiting for all Antrea DaemonSet Pods
##[error]    networkpolicy_test.go:89: Pod test-pod-networkpolicy-kii1a86r should not be able to connect Service test-svc-nginx-r2w1u3or, but was able to connect
##[error]    fixtures.go:185: Deleting Pod 'test-pod-networkpolicy-3lw8juxq'
##[error]    fixtures.go:185: Deleting Pod 'test-pod-networkpolicy-kii1a86r'
##[error]    fixtures.go:185: Deleting Pod 'test-pod-nginx-b95n6a8w'
##[error]    fixtures.go:71: Exporting test logs to '/tmp/antrea-test-163908562/TestIPBlockWithExcept.Dec17-19-31-47'
##[error]    fixtures.go:178: Deleting 'antrea-test' K8s Namespace
FAIL
2019/12/17 19:32:16 Logs exported under '/tmp/antrea-test-163908562', it is your responsibility to delete the directory when you no longer need it
FAIL	github.com/vmware-tanzu/antrea/test/e2e	321.599s
FAIL
```
Is it possible that the deny-all network policy takes too long to be enforced for the new Pods? It is a bit unlikely because we actually wait for the Pods' IPs to become available before running the `wget` command, but still maybe it is worth waiting for https://github.com/vmware-tanzu/antrea/issues/197 to be merged before investigating in details.

Have met similar issue but got error from a different line.
` ##[error]    networkpolicy_test.go:108: Pod test-pod-networkpolicy-f0sxvq4a should be able to connect Service test-svc-nginx-ee8igovv, but was not able to connect`

--- a/docs/kind.md
+++ b/docs/kind.md
@@ -56,17 +56,24 @@ antrea-agent-zsztq                   2/2     Running   0          8m56s
 antrea-controller-775f4d79f8-6tksp   1/1     Running   0          8m56s
 ```
 
-### Short Cut
-Alternatively to create a two worker Node cluster with antrea installed, do
+### Short-Cut
+
+Alternatively to create a two worker Node cluster with Antrea installed, do
 ```
 ./ci/kind/kind-setup.sh create CLUSTER_NAME
 ```
-kind-setup.sh allows user to specify number of worker Nodes, docker bridge networks/subnets
-connected to worker Nodes, and preloaded docker images. For their usages do 
+kind-setup.sh allows users to specify the number of worker Nodes, the docker
+bridge networks/subnets connected to worker Nodes, and some docker images to be
+pre-loaded in each Node. For more information on usage, run:
  ```
 ./ci/kind/kind-setup.sh help
 ```
  
+## Run the Antrea e2e tests
+
+To run the Antrea e2e test suite on your Kind cluster, please refer to [this
+document](/test/e2e#running-the-e2e-tests-on-a-kind-cluster).
+
 ## FAQ
 
 ### Why is the YAML manifest different when using Kind?

--- a/test/e2e/connectivity_test.go
+++ b/test/e2e/connectivity_test.go
@@ -164,17 +164,17 @@ func (data *TestData) redeployAntrea(t *testing.T, enableIPSec bool) {
 	if err := data.waitForAntreaDaemonSetPods(defaultTimeout); err != nil {
 		t.Fatalf("Error when restarting Antrea: %v", err)
 	}
-	t.Logf("Checking CoreDNS deployment")
-	if err := data.checkCoreDNSPods(defaultTimeout); err != nil {
-		t.Fatalf("Error when checking CoreDNS deployment: %v", err)
+	// Restart CoreDNS Pods to avoid issues caused by disrupting the datapath (when restarting
+	// Antrea Agent Pods).
+	t.Logf("Restarting CoreDNS Pods")
+	if err := data.restartCoreDNSPods(defaultTimeout); err != nil {
+		t.Fatalf("Error when restarting CoreDNS Pods: %v", err)
 	}
 }
 
 // TestPodConnectivityAfterAntreaRestart checks that restarting antrea-agent does not create
 // connectivity issues between Pods.
 func TestPodConnectivityAfterAntreaRestart(t *testing.T) {
-	// See https://github.com/vmware-tanzu/antrea/issues/244
-	skipIfProviderIs(t, "kind", "test may cause subsequent tests to fail in Kind clusters")
 	data, err := setupTest(t)
 	if err != nil {
 		t.Fatalf("Error when setting up test: %v", err)

--- a/test/e2e/framework.go
+++ b/test/e2e/framework.go
@@ -286,7 +286,7 @@ func (data *TestData) deployAntreaIPSec() error {
 // waitForAntreaDaemonSetPods waits for the K8s apiserver to report that all the Antrea Pods are
 // available, i.e. all the Nodes have one or more of the Antrea daemon Pod running and available.
 func (data *TestData) waitForAntreaDaemonSetPods(timeout time.Duration) error {
-	err := wait.Poll(1*time.Second, timeout, func() (bool, error) {
+	err := wait.PollImmediate(1*time.Second, timeout, func() (bool, error) {
 		daemonSet, err := data.clientset.AppsV1().DaemonSets(antreaNamespace).Get(antreaDaemonSet, metav1.GetOptions{})
 		if err != nil {
 			return false, fmt.Errorf("error when getting Antrea daemonset: %v", err)
@@ -308,28 +308,9 @@ func (data *TestData) waitForAntreaDaemonSetPods(timeout time.Duration) error {
 	return nil
 }
 
-// checkCoreDNSPods checks that all the Pods for the CoreDNS deployment are ready. If not, delete
-// all the Pods to force them to restart and waits up to timeout for the Pods to become ready.
-func (data *TestData) checkCoreDNSPods(timeout time.Duration) error {
-	if deployment, err := data.clientset.AppsV1().Deployments(antreaNamespace).Get("coredns", metav1.GetOptions{}); err != nil {
-		return fmt.Errorf("error when retrieving CoreDNS deployment: %v", err)
-	} else if deployment.Status.UnavailableReplicas == 0 {
-		// deployment ready, nothing to do
-		return nil
-	}
-
-	// restart CoreDNS and wait for all replicas
-	var gracePeriodSeconds int64 = 1
-	deleteOptions := &metav1.DeleteOptions{
-		GracePeriodSeconds: &gracePeriodSeconds,
-	}
-	listOptions := metav1.ListOptions{
-		LabelSelector: "k8s-app=kube-dns",
-	}
-	if err := data.clientset.CoreV1().Pods(antreaNamespace).DeleteCollection(deleteOptions, listOptions); err != nil {
-		return fmt.Errorf("error when deleting all CoreDNS Pods: %v", err)
-	}
-	err := wait.Poll(1*time.Second, timeout, func() (bool, error) {
+// waitForCoreDNSPods waits for the K8s apiserver to report that all the CoreDNS Pods are available.
+func (data *TestData) waitForCoreDNSPods(timeout time.Duration) error {
+	err := wait.PollImmediate(1*time.Second, timeout, func() (bool, error) {
 		deployment, err := data.clientset.AppsV1().Deployments(antreaNamespace).Get("coredns", metav1.GetOptions{})
 		if err != nil {
 			return false, fmt.Errorf("error when retrieving CoreDNS deployment: %v", err)
@@ -346,6 +327,35 @@ func (data *TestData) checkCoreDNSPods(timeout time.Duration) error {
 		return err
 	}
 	return nil
+}
+
+// restartCoreDNSPods deletes all the CoreDNS Pods to force them to be re-scheduled. It then waits
+// for all the Pods to become available, by calling waitForCoreDNSPods.
+func (data *TestData) restartCoreDNSPods(timeout time.Duration) error {
+	var gracePeriodSeconds int64 = 1
+	deleteOptions := &metav1.DeleteOptions{
+		GracePeriodSeconds: &gracePeriodSeconds,
+	}
+	listOptions := metav1.ListOptions{
+		LabelSelector: "k8s-app=kube-dns",
+	}
+	if err := data.clientset.CoreV1().Pods(antreaNamespace).DeleteCollection(deleteOptions, listOptions); err != nil {
+		return fmt.Errorf("error when deleting all CoreDNS Pods: %v", err)
+	}
+	return data.waitForCoreDNSPods(timeout)
+}
+
+// checkCoreDNSPods checks that all the Pods for the CoreDNS deployment are ready. If not, it
+// deletes all the Pods to force them to restart and waits up to timeout for the Pods to become
+// ready.
+func (data *TestData) checkCoreDNSPods(timeout time.Duration) error {
+	if deployment, err := data.clientset.AppsV1().Deployments(antreaNamespace).Get("coredns", metav1.GetOptions{}); err != nil {
+		return fmt.Errorf("error when retrieving CoreDNS deployment: %v", err)
+	} else if deployment.Status.UnavailableReplicas == 0 {
+		// deployment ready, nothing to do
+		return nil
+	}
+	return data.restartCoreDNSPods(timeout)
 }
 
 // createClient initializes the K8s clientset in the TestData structure.

--- a/test/e2e/providers/kind.go
+++ b/test/e2e/providers/kind.go
@@ -22,6 +22,9 @@ import (
 	"github.com/vmware-tanzu/antrea/test/e2e/providers/exec"
 )
 
+// TODO: try not to hardcode this name if possible.
+const masterNodeName = "kind-control-plane"
+
 type KindProvider struct{}
 
 func (provider *KindProvider) RunCommandOnNode(nodeName string, cmd string) (
@@ -45,11 +48,29 @@ func (provider *KindProvider) GetKubeconfigPath() (string, error) {
 // enableKubectlOnMaster copies the Kubeconfig file on the Kind control-plane / master Node to the
 // default location, in order to make sure that we can run kubectl on the Node.
 func (provider *KindProvider) enableKubectlOnMaster() error {
-	// TODO: try not to hardcode this name if possible.
-	nodeName := "kind-control-plane"
-	rc, stdout, _, err := provider.RunCommandOnNode(nodeName, "cp /etc/kubernetes/admin.conf /root/.kube/config")
+	rc, stdout, _, err := provider.RunCommandOnNode(masterNodeName, "cp /etc/kubernetes/admin.conf /root/.kube/config")
 	if err != nil || rc != 0 {
-		return fmt.Errorf("error when copying Kubeconfig file to /root/.kube/config on '%s': %s", nodeName, stdout)
+		return fmt.Errorf("error when copying Kubeconfig file to /root/.kube/config on '%s': %s", masterNodeName, stdout)
+	}
+	return nil
+}
+
+// moveCoreDNSPodsToMaster ensures that all the CoreDNS Pods are scheduled on master by patching the
+// CoreDNS deployment with kubectl. Several of the e2e tests restart the Antrea agent on a worker
+// Node, which causes the datapath to break. If this happens when CoreDNS Pods are scheduled on this
+// specific worker Node, we observe that kube-apiserver cannot reach CoreDNS any more, which may
+// cause some subsequent tests to fail. Because we never restart the Antrea agent on the master
+// Node, it helps to move the CoreDNS Pods to master. Note that CoreDNS Pods still need to be
+// restarted if a test restarts all Agent Pods (e.g. as part of a rolling update when the Antrea
+// YAML manifest is changed). This issue does not seem to affect clusters which use the OVS kernel
+// datapath as much.
+// TODO: revert changes at the end of tests?
+func (provider *KindProvider) moveCoreDNSPodsToMaster() error {
+	patch := `{"spec":{"template":{"spec":{"nodeName":"kind-control-plane"}}}}`
+	cmd := fmt.Sprintf("kubectl patch -v 8 deployment coredns -n kube-system -p %s", patch)
+	rc, stdout, _, err := provider.RunCommandOnNode(masterNodeName, cmd)
+	if err != nil || rc != 0 {
+		return fmt.Errorf("error when scheduling CoreDNS Pods to '%s': %s", masterNodeName, stdout)
 	}
 	return nil
 }
@@ -60,6 +81,9 @@ func (provider *KindProvider) enableKubectlOnMaster() error {
 func NewKindProvider(configPath string) (ProviderInterface, error) {
 	provider := &KindProvider{}
 	if err := provider.enableKubectlOnMaster(); err != nil {
+		return nil, err
+	}
+	if err := provider.moveCoreDNSPodsToMaster(); err != nil {
 		return nil, err
 	}
 	return provider, nil

